src/Mod/Mesh/App/Exporter.cpp

```diff
@@ -0,0 +1,391 @@
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+    #include <algorithm>
+    #include <vector>
+    #include <boost/algorithm/string/replace.hpp>
+#endif  //  #ifndef _PreComp_
+
+#include "Exporter.h"
+
+#include "Core/Iterator.h"
+
+#include "Base/Console.h"
+#include "Base/Exception.h"
+#include "Base/FileInfo.h"
+#include "Base/Sequencer.h"
+#include "Base/Stream.h"
+#include "Base/Tools.h"
+
+#include "App/Part.h"
+
+#include <zipios++/zipoutputstream.h>
+
+using namespace Mesh;
+using namespace MeshCore;
+
+Exporter::Exporter() :
+    meshFeatId( Base::Type::fromName("Mesh::Feature") ),
+    appPartId( Base::Type::fromName("Part::Feature") ),
+    groupExtensionId( App::GroupExtension::getExtensionClassTypeId() )
+{ }
+
+//static
+std::string Exporter::xmlEscape(const std::string &input)
+{
+    std::string out(input);
+    boost::replace_all(out, "&", "&amp;");
+    boost::replace_all(out, "\"", "&quot;");
+    boost::replace_all(out, "'", "&apos;");
+    boost::replace_all(out, "<", "&lt;");
+    boost::replace_all(out, ">", "&gt;");
+    return out;
+}
+
+bool Exporter::addAppGroup(App::DocumentObject *obj, float tol)
+{
+    auto ret(true);
+
+    for (auto it : static_cast<App::Part *>(obj)->getOutList()) {
+        if (it->getTypeId().isDerivedFrom(meshFeatId)) {
+            ret &= addMeshFeat(it);
+        } else if (it->getTypeId().isDerivedFrom(appPartId)) {
+            ret &= addPartFeat(it, tol);
+        } else if (it->hasExtension(groupExtensionId)) {
+            // Recurse
+            ret &= addAppGroup(it, tol);
+        }
+    }
+
+    return ret;
+}
+
+bool Exporter::addObject(App::DocumentObject *obj, float tol)
+{
+    if (obj->getTypeId().isDerivedFrom(meshFeatId)) {
+        return addMeshFeat( obj );
+    } else if (obj->getTypeId().isDerivedFrom(appPartId)) {
+        return addPartFeat( obj, tol );
+    } else if (obj->hasExtension(groupExtensionId)) {
```

Comment:
@ickby - I've only briefly tested this (late here), but is this more like you would expect?  The idea is that whatever is selected by the user is passed in to addObject, and any meshable objects within should end up being added by either addMeshFeat() or addPartFeat().